<a href="https://colab.research.google.com/github/skarot/DeepLearning/blob/main/FashionMnistANN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Image classification using ANN on fashion-mnist dataset

## Import libraries

In [1]:
import tensorflow as tf
from tensorflow import keras
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


## Load data

In [2]:
(X_train,y_train),(X_test,y_test) = keras.datasets.fashion_mnist.load_data()

4431872/4422102 [==============================] - 0s 0us/step


## Data Analysis

In [17]:
X_train[0]


array([[  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   1,
          0,   0,  13,  73,   0,   0,   1,   4,   0,   0,   0,   0,   1,
          1,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   3,
          0,  36, 136, 127,  62,  54,   0,   0,   0,   1,   3,   4,   0,
          0,   3],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   6,
          0, 102, 204, 176, 134, 144, 123,  23,   0,   0,   0,   0,  12,
         10,   0],
       [  

In [6]:
X_train.shape

(60000, 28, 28)

In [7]:
y_train

array([9, 0, 0, ..., 3, 0, 5], dtype=uint8)

In [8]:
y_train.shape

(60000,)

In [9]:
y_train.dtype

dtype('uint8')

In [13]:
np.unique(y_train)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], dtype=uint8)

In [15]:
y_train[0:4]

array([9, 0, 0, 3], dtype=uint8)

y_train has descretet values hence we will be using sparse_categorical_crossentropy as loss

In [14]:
classes = ['T-shirt/top','Trouser','Pullover','Dress','Coat','Sandal','Shirt','Sneaker','Bag','Ankle boot']

## Scaling 

In [20]:
X_train_scaled = X_train/255
X_test_scaled = X_test/255

In [ ]:
X_train_scaled[0]

In [23]:
y_train_scaled

array([0.9, 0. , 0. , ..., 0.3, 0. , 0.5])

## Model Creation

In [27]:
model = keras.Sequential([
                          keras.layers.Flatten(input_shape = (28,28)),
                          keras.layers.Dense(100,activation='relu'),
                          keras.layers.Dense(10,activation= 'sigmoid')
])

model.compile(optimizer='adam',loss='sparse_categorical_crossentropy', metrics = 'accuracy')

## Training

In [29]:
model.fit(X_train_scaled,y_train,epochs= 50)

Epoch 1/50
1875/1875 [==============================] - 5s 2ms/step - loss: 1.1500 - accuracy: 0.7986
Epoch 2/50
1875/1875 [==============================] - 5s 2ms/step - loss: 0.4176 - accuracy: 0.8536
Epoch 3/50
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3883 - accuracy: 0.8621
Epoch 4/50
1875/1875 [==============================] - 5s 2ms/step - loss: 0.3701 - accuracy: 0.8678
Epoch 5/50
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3551 - accuracy: 0.8738
Epoch 6/50
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3432 - accuracy: 0.8777
Epoch 7/50
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3368 - accuracy: 0.8800
Epoch 8/50
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3281 - accuracy: 0.8811
Epoch 9/50
1875/1875 [==============================] - 5s 2ms/step - loss: 0.3213 - accuracy: 0.8847
Epoch 10/50
1875/1875 [==============================] - 5s 3ms/step - loss: 0.315

## Predict

In [30]:
y_hat = model.predict(X_test)

## Evaluate

In [31]:
loss, accuracy = model.evaluate(X_test_scaled,y_test)

313/313 [==============================] - 1s 2ms/step - loss: 0.4136 - accuracy: 0.8718


## Test

Original class

In [40]:
y_test[2]

1

In [37]:
classes[y_test[2]]

'Trouser'

Predicted Class

In [39]:
y_hat[2]

array([0., 1., 0., 0., 1., 0., 0., 0., 0., 0.], dtype=float32)

In [38]:
classes[np.argmax(y_hat[2])]

'Trouser'

## Test with hiddenlayers increased to 5

A function to create hidden layers dynamically based on no of hidden lyers as input

In [47]:
def get_hidden_layer_list(hidden_layers):
  ls = []
  ls.append(keras.layers.Flatten(input_shape = (28,28)))
  for i in range(1,hidden_layers+1):
    ls.append(keras.layers.Dense((700/i), activation='relu'))
  ls.append(keras.layers.Dense(10,activation = 'sigmoid'))
  return ls


Common function to get model based on CPU or GPU device setting

In [48]:
def get_model(hidden_layers):  
  ls = get_hidden_layer_list(hidden_layers)
  model = keras.Sequential(ls)

  return model

In [64]:
%%timeit -n1 -r1
with tf.device('/CPU:0'):
  model = get_model(1)
  model.compile(optimizer='adam',loss ='sparse_categorical_crossentropy', metrics = 'accuracy')
  model.fit(X_train_scaled,y_train, epochs = 1)


1875/1875 [==============================] - 8s 4ms/step - loss: 0.4733 - accuracy: 0.8301
1 loop, best of 1: 8.19 s per loop


In [65]:
model.evaluate(X_test_scaled,y_test)

313/313 [==============================] - 1s 2ms/step - loss: 0.3429 - accuracy: 0.8844


[0.3428660035133362, 0.8844000101089478]

In [66]:
%%timeit -n1 -r1
with tf.device('/GPU:0'):
  model = get_model(1)
  model.compile(optimizer='adam',loss ='sparse_categorical_crossentropy', metrics = 'accuracy')
  model.fit(X_train_scaled,y_train, epochs = 1)


1875/1875 [==============================] - 4s 2ms/step - loss: 0.4734 - accuracy: 0.8301
1 loop, best of 1: 4.39 s per loop


In [67]:
model.evaluate(X_test_scaled,y_test)

313/313 [==============================] - 1s 2ms/step - loss: 0.3429 - accuracy: 0.8844


[0.3428660035133362, 0.8844000101089478]

In [68]:
%%timeit -n1 -r1
with tf.device('/CPU:0'):
  model = get_model(5)
  model.compile(optimizer='adam',loss ='sparse_categorical_crossentropy', metrics = 'accuracy')
  model.fit(X_train_scaled,y_train, epochs = 10)


Epoch 1/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.5015 - accuracy: 0.8185
Epoch 2/10
1875/1875 [==============================] - 14s 7ms/step - loss: 0.3805 - accuracy: 0.8608
Epoch 3/10
1875/1875 [==============================] - 15s 8ms/step - loss: 0.3386 - accuracy: 0.8764
Epoch 4/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.3195 - accuracy: 0.8837
Epoch 5/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.2993 - accuracy: 0.8912
Epoch 6/10
1875/1875 [==============================] - 14s 7ms/step - loss: 0.2842 - accuracy: 0.8953
Epoch 7/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.2742 - accuracy: 0.8986
Epoch 8/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.2610 - accuracy: 0.9025
Epoch 9/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.2522 - accuracy: 0.9056
Epoch 10/10
1875/1875 [==============================] - 13s 7ms/step - l

In [69]:
model.evaluate(X_test_scaled,y_test)

313/313 [==============================] - 1s 2ms/step - loss: 0.3429 - accuracy: 0.8844


[0.3428660035133362, 0.8844000101089478]

In [70]:
%%timeit -n1 -r1
with tf.device('/GPU:0'):
  model = get_model(5)
  model.compile(optimizer='adam',loss ='sparse_categorical_crossentropy', metrics = 'accuracy')
  model.fit(X_train_scaled,y_train, epochs = 10)


Epoch 1/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.5067 - accuracy: 0.8154
Epoch 2/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.3833 - accuracy: 0.8609
Epoch 3/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.3452 - accuracy: 0.8751
Epoch 4/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.3167 - accuracy: 0.8849
Epoch 5/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.2990 - accuracy: 0.8910
Epoch 6/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.2874 - accuracy: 0.8938
Epoch 7/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.2728 - accuracy: 0.8988
Epoch 8/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.2606 - accuracy: 0.9018
Epoch 9/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.2515 - accuracy: 0.9073
Epoch 10/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.242

In [71]:
model.evaluate(X_test_scaled,y_test)

313/313 [==============================] - 1s 2ms/step - loss: 0.3429 - accuracy: 0.8844


[0.3428660035133362, 0.8844000101089478]

Final testing

In [75]:
classes[y_test[5]]


'Trouser'

In [74]:

classes[np.argmax(y_hat[5])]

'Trouser'

## Final performance report

Hidden Layers | Epochs | CPU      | GPU    |
--------------|--------|----------|--------|
1             |   1    | 8.19 s   |4.39 s  |
5             |   10   | 2min 11s |1min 22s|